Add support for Detection task type

djdameln opened the pull request:
# Description

This PR adds the detection task type. This is mainly to support region-based models that produce bounding box predictions instead of / in addition to anomaly maps.

## Changes
### Major
- Adds conversion utilities between masks and bounding boxes. This allows automatic generation of ground truth boxes for detection models. These utilities are also used to generate bounding box predictions when the model is of the segmentation type.
- The detection task uses the thresholding and evaluation mechanisms from the segmentation task under the hood. This is achieved by converting the bounding boxes to segmentation masks or anomaly maps before evaluating the metrics.
- Visualization functionality was added. This means that the predictions of any model capable of localization (segmentation or detection) can be visualized as detection predictions. For example, the following image was generated by passing `detection` as task type for a padim model training on MVTec bottle dataset:

![002](https://user-images.githubusercontent.com/26067904/203837192-a2392d63-f911-4b0a-9a59-e76d0280ded9.png)

### Minor
- Using enum instead of string to configure task type.
- A new `collate_fn` function had to be implemented to allow collating the ground truth bounding boxes across batches.

## Known issues
- No test cases (WIP)
- No inference support

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://github.com/openvinotoolkit/anomalib/blob/main/CONTRIBUTING.md) of this project.
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

samet-akcay commented on anomalib/data/__init__.py:
```diff
@@ -14,6 +14,7 @@
 from .folder import Folder
 from .inference import InferenceDataset
 from .mvtec import MVTec
+from .task_type import TaskType
```
would this be used by models and other components as well? If so, would it be an idea to move it to a place that would be accessible by other components?

djdameln replied:
Yeah, I'm not super happy with the location of `TaskType` but couldn't think of a better place. Maybe somewhere in `anomalib/config` would work? @ashwinvaidya17 any ideas?

ashwinvaidya17 reviewed (commented):
Thanks for the efforts! I have a few minor comments

samet-akcay commented on anomalib/data/base/datamodule.py:
```diff
@@ -12,14 +12,28 @@
 
+def collate_fn(batch):
+    """Custom collate function that collates bounding boxes as lists."""
+    elem = batch[0]
```
What does `elem` mean, element? Would it be possible to rename the variable to show the intention?
Also what else the type would be? Would it be possible to annotate it?

djdameln replied:
elem stands for 'element'. The naming was chosen to match the naming used in [torch's default collate function](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/_utils/collate.py#L115). A single element is taken from the batch to check if the batch is a list of dictionaries, and to retrieve the keys of the dictionary that need to be collated.

I'll add an inline comment to clarify this.

samet-akcay commented on anomalib/data/base/datamodule.py:
```diff
@@ -101,8 +115,20 @@ def train_dataloader(self) -> TRAIN_DATALOADERS:
     def val_dataloader(self) -> EVAL_DATALOADERS:
         """Get validation dataloader."""
-        return DataLoader(self.val_data, shuffle=False, batch_size=self.eval_batch_size, num_workers=self.num_workers)
+        return DataLoader(
+            self.val_data,
+            shuffle=False,
+            batch_size=self.eval_batch_size,
+            num_workers=self.num_workers,
+            collate_fn=collate_fn,
+        )
```
Sorry, it's a bit picky comment, but would be be easier to read if all the arguments are specified.
```suggestion
        return DataLoader(
            dataset=self.val_data,
            shuffle=False,
            batch_size=self.eval_batch_size,
            num_workers=self.num_workers,
            collate_fn=collate_fn,
        )
```

samet-akcay commented on anomalib/data/base/datamodule.py:
```diff
@@ -101,8 +115,20 @@ def train_dataloader(self) -> TRAIN_DATALOADERS:
     def test_dataloader(self) -> EVAL_DATALOADERS:
         """Get test dataloader."""
-        return DataLoader(self.test_data, shuffle=False, batch_size=self.eval_batch_size, num_workers=self.num_workers)
+        return DataLoader(
+            self.test_data,
+            shuffle=False,
+            batch_size=self.eval_batch_size,
+            num_workers=self.num_workers,
+            collate_fn=collate_fn,
+        )
```
```suggestion
        return DataLoader(
            dataset=self.test_data,
            shuffle=False,
            batch_size=self.eval_batch_size,
            num_workers=self.num_workers,
            collate_fn=collate_fn,
        )
```

samet-akcay commented on anomalib/data/base/video.py:
```diff
@@ -74,9 +78,15 @@ def __getitem__(self, index: int) -> Dict[str, Union[str, Tensor]]:
+            mask = Tensor(item["mask"])
             item["mask"] = torch.stack([item["mask"] for item in processed_frames]).squeeze(0)
             item["label"] = Tensor([1 in frame for frame in mask]).int().squeeze(0)
+            if self.task == TaskType.DETECTION:
+                item["boxes"] = [
+                    torch.empty((0, 4)) if frame.max() == 0 else masks_to_boxes(frame)
+                    for frame in item["mask"].view((-1, 1) + item["mask"].shape[-2:])
+                ]
```
This line is a bit difficult to follow. Would it be possible to have a specific function for this line so the reader could go over it line by line for clarity?

djdameln replied:
It turns out this was unnecessarily complicated and could be reduced to a single assignment.

samet-akcay commented on anomalib/data/folder.py:
```diff
@@ -141,7 +142,7 @@ class FolderDataset(AnomalibDataset):
 
     Args:
-        task (str): Task type. (classification or segmentation).
+        task (TaskType): Task type. (classification, detection or segmentation).
```
For consistency, following your changes above.
```suggestion
        task (TaskType): Task type. (``classification``, ``detection`` or ``segmentation``).
```

samet-akcay commented on anomalib/data/folder.py:
```diff
@@ -222,8 +223,8 @@ class Folder(AnomalibDataModule):
         num_workers (int, optional): Number of workers. Defaults to 8.
-        task (str, optional): Task type. Could be either classification or segmentation.
-            Defaults to "classification".
+        task (TaskType, optional): Task type. Could be classification, detection or segmentation.
+            Defaults to segmentation.
```
```suggestion
        task (TaskType, optional): Task type. Could be ``classification``, ``detection`` or ``segmentation``.
            Defaults to ``segmentation``.
```

samet-akcay commented on anomalib/data/mvtec.py:
```diff
@@ -123,7 +124,7 @@ class MVTecDataset(AnomalibDataset):
 
     Args:
-        task (str): Task type, either 'classification' or 'segmentation'
+        task (TaskType): Task type,'classification', 'detection' or 'segmentation'
```
```suggestion
        task (TaskType): Task type, ``classification``,``detection`` or ``segmentation``
```

samet-akcay commented on anomalib/data/task_type.py:
```diff
@@ -0,0 +1,11 @@
+"""Task type enum."""
+
```
Would be good to add the license.

samet-akcay commented on anomalib/data/utils/boxes.py:
```diff
@@ -0,0 +1,81 @@
+"""Helper functions for processing bounding box detections and annotations."""
+
```
License is missing

samet-akcay commented on anomalib/data/utils/boxes.py:
```diff
@@ -0,0 +1,81 @@
+    masks = masks.view((-1, 1) + masks.shape[-2:])  # reshape to (B, 1, H, W)
+    masks = masks.float()
+
+    if masks.is_cuda:
+        batch_comps = connected_components_gpu(masks).squeeze(1)
+    else:
+        batch_comps = connected_components_cpu(masks).squeeze(1)
```
A bit opinionated, but how about this? You don't need to accept this change.
```suggestion
    connected_components = connected_components_gpu if masks.is_cuda else connected_components_cpu
    batch_comps = connected_components(masks).squeeze(1)
```

djdameln replied:
mypy was causing some trouble after I committed this change, so I reverted it.

samet-akcay commented on anomalib/deploy/inferencers/openvino_inferencer.py:
```diff
@@ -159,7 +160,7 @@ def post_process(
             pred_label = pred_score >= meta_data["image_threshold"]
 
-        if task == "segmentation":
+        if task == TaskType.SEGMENTATION:
```
Do we have any plans on supporting detection in OpenVINO inference?

djdameln replied:
Inference support will follow. I don't see any problems with adding OpenVINO support, but will need to check to confirm.

samet-akcay commented on anomalib/models/components/base/anomaly_module.py:
```diff
@@ -82,6 +84,8 @@ def predict_step(self, batch: Any, batch_idx: int, _dataloader_idx: Optional[int
         outputs["pred_labels"] = outputs["pred_scores"] >= self.image_threshold.value
         if "anomaly_maps" in outputs.keys():
             outputs["pred_masks"] = outputs["anomaly_maps"] >= self.pixel_threshold.value
+            if "pred_boxes" not in outputs.keys():
+                outputs["pred_boxes"] = masks_to_boxes(outputs["pred_masks"])
```
Is it possible to add `pred_boxes` to `outputs` only when the task type is `detection`? Or am I missing something?

djdameln replied:
The problem is that we can't access task type information from within the lightning module. This is one of the reasons why we should consider switching to custom training loops.

samet-akcay commented on anomalib/models/components/base/anomaly_module.py:
```diff
@@ -155,15 +159,28 @@ def _collect_outputs(image_metric, pixel_metric, outputs):
+        elif "pred_scores" not in outputs and "boxes_scores" in outputs:
+            # infer image score from bbox confidence scores
+            outputs["pred_scores"] = torch.stack([scores.max() for scores in outputs["boxes_scores"]])
+        if "pred_boxes" in outputs and "anomaly_maps" not in outputs:
```
Can we have a new line above as it initially looks confusing to see an `if ` statement right after `elif`

samet-akcay commented on anomalib/utils/cv/connected_components.py:
```diff
@@ -0,0 +1,48 @@
+"""Connected component labeling."""
+
```
License?

samet-akcay reviewed (commented):
Took me some time to review :) Appreciate your time to finish the entire implementation :)

jpcbertoldo commented:
> Using enum instead of string to configure task type.

I found a few str <--> enum conversion issues at some point when trying to use the cli with config files (somehow the string from the `.yaml` was not properly converted?).

So it made me wonder: are there tests covering this? (i.e. the different values of the enum work properly when being passed from the config file)

djdameln commented on anomalib/utils/callbacks/min_max_normalization.py:
```diff
@@ -46,10 +47,14 @@ def on_validation_batch_end(
+        elif "pred_scores" in outputs:
             pl_module.normalization_metrics(outputs["pred_scores"])
+        else:
+            raise ValueError("No values found for normalization, provide anomaly maps, bbox scores, or image scores")
```
I agree that `image_score` would be more suitable, but I feel changing this is outside of the scope of this PR.

djdameln commented:
> So it made me wonder: are there tests covering this? (i.e. the different values of the enum work properly when being passed from the config file)

I don't think the tests cover this explicitly at the moment. Which enum was causing problems at your end?

jpcbertoldo commented:
> > So it made me wonder: are there tests covering this? (i.e. the different values of the enum work properly when being passed from the config file)
> 
> I don't think the tests cover this explicitly at the moment. Which enum was causing problems at your end?

Honestly I don't remember because it was blocking me to use the new cli so I dropped the ball.
Anyway, just something to keep in mind : )

djdameln commented:
@ashwinvaidya17 @samet-akcay I've updated the PR based on your comments, and also added test cases and inference support. So it's ready for the next round of review.

ashwinvaidya17 commented on anomalib/utils/callbacks/visualizer/visualizer_base.py:
```diff
@@ -37,8 +38,10 @@ def __init__(
-            raise ValueError(f"Unknown task type: {mode}. Please choose one of ['classification', 'segmentation']")
+        if task not in [TaskType.CLASSIFICATION, TaskType.DETECTION, TaskType.SEGMENTATION]:
+            raise ValueError(
+                f"Unknown task type: {mode}. Please choose one of ['classification', 'detection', 'segmentation']"
```
Not really necessary here but I wonder if `f"Unknown task type: {mode}. Please choose one of {TaskType.__members__.values()}` will work here. But will not make it a very descriptive docstring for the programmers.

ashwinvaidya17 reviewed (approved):
Thanks for the efforts!

samet-akcay reviewed (approved):
LGTM, thanks!
